ecosystem/gstreamer_plugin/gst_mtl_st40_rx.c

```diff
@@ -0,0 +1,509 @@
+    gst_mtl_common_set_general_argumetns(object, prop_id, value, pspec, &(self->devArgs),
+                                         &(self->portArgs), &self->log_level);
+    return;
+  }
+
+  switch (prop_id) {
+    case PROP_ST40_RX_BUFFER_SIZE:
+      self->mbuff_size = g_value_get_uint(value);
+      break;
+    case PROP_ST40_RX_TIMEOUT_MBUF_GET:
+      self->timeout_mbuf_get_seconds = g_value_get_uint(value);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+      break;
+  }
+}
+
+static void gst_mtl_st40_rx_get_property(GObject* object, guint prop_id, GValue* value,
+                                         GParamSpec* pspec) {
+  Gst_Mtl_St40_Rx* src = GST_MTL_ST40_RX(object);
+
+  if (prop_id < PROP_GENERAL_MAX) {
+    gst_mtl_common_get_general_argumetns(object, prop_id, value, pspec, &(src->devArgs),
+                                         &(src->portArgs), &src->log_level);
+    return;
+  }
+
+  switch (prop_id) {
+    case PROP_ST40_RX_BUFFER_SIZE:
+      g_value_set_uint(value, src->mbuff_size);
+      break;
+    case PROP_ST40_RX_TIMEOUT_MBUF_GET:
+      g_value_set_uint(value, src->timeout_mbuf_get_seconds);
+      break;
+    default:
+      G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+      break;
+  }
+}
+
+static void* gst_mtl_st40_rx_get_mbuf_with_timeout(Gst_Mtl_St40_Rx* src,
+                                                   st40_rx_handle handle, void** usrptr,
+                                                   uint16_t* size) {
+  struct timespec ts;
+  gint ret;
+  void* mbuf = NULL;
+
+  clock_gettime(CLOCK_REALTIME, &ts);
+  ts.tv_sec += src->timeout_mbuf_get_seconds;
+  ret = pthread_mutex_timedlock(&(src->get_mbuff_mutex), &ts);
+
+  if (!ret)
+    ret = pthread_cond_timedwait(&(src->get_mbuff_cond), &(src->get_mbuff_mutex), &ts);
+
+  if (ret == ETIMEDOUT) return NULL;
+
+  pthread_mutex_unlock(&(src->get_mbuff_mutex));
+
+  mbuf = st40_rx_get_mbuf(src->rx_handle, usrptr, size);
+  if (!mbuf) GST_ERROR("Failed to get ancillary mbuf\n");
+
+  return mbuf;
+}
+
+static GstFlowReturn gst_mtl_st40_rx_fill_buffer(Gst_Mtl_St40_Rx* src, GstBuffer** buffer,
+                                                 void* usrptr) {
+  struct st40_rfc8331_rtp_hdr* hdr;
+  struct st40_rfc8331_payload_hdr* payload_hdr;
+  GstMapInfo dest_info;
+  guint16 data, fill_size;
+  gint udw_size;
+
+  hdr = (struct st40_rfc8331_rtp_hdr*)usrptr;
+  payload_hdr = (struct st40_rfc8331_payload_hdr*)(&hdr[1]);
+  udw_size = payload_hdr->second_hdr_chunk.data_count & 0xff;
+
+  if (src->udw_size == 0) {
+    src->udw_size = udw_size;
+    src->anc_data = (char*)malloc(udw_size);
+  } else if (src->udw_size != udw_size) {
+    GST_WARNING("Size of recieved ancillary data has changed");
+    if (src->anc_data) {
+      free(src->anc_data);
+      src->anc_data = NULL;
+    }
+    src->udw_size = udw_size;
+    src->anc_data = (char*)malloc(udw_size);
+  }
+
+  if (udw_size == 0) {
```

Comment:
I guess it should be higher as this value is assigned to other variables in previous `if`.

Reply:
moved to the top